### PR TITLE
Implement Sign & Verify

### LINKED ISSUE #146: Sign and Verify messages
### Description of the Feature or Idea

Add the ability for websites to request the visitor's LN node to sign or verify a message according to the WebLN provider spec https://github.com/wbobeirne/webln#provider-spec.

```ts
  /* Prompts the user to sign a message with their private key */
  signMessage(message: string): Promise<SignMessageResponse>;

  /* Shows the user a view that verifies a signed message */
  verifyMessage(signedMessage: string, rawMessage: string): Promise<void>;
``` 

### Limitations

LND requires nodes be routable between each other in order to verify a message. In other words, given the lightning networks `A -> B -> C` & `D <- E`. If `A` signs a message, then `B` and `C` are able to verify that message but `D` and `E` are not.

The `lncli verifymessage` command returns `valid: false` in these two cases:
- the message and signature do not match
- the signing node is not reachable from the verifying node on the network 

Joule will not be able to inform the website if the failure is one or the other.

### Use Cases

**Sign**
The obvious use case is for authentication. The website can give the user a message to sign using Joule then take the signature and verify it on the backend to confirm the identity of the user's node. 

**Verify**
??? 🤔 

I've started implementing this functionality. I'm creating this PR to open the floor for suggestions or ideas for improvements

### PR DESCRIPTION
**Pull request by jamaljsr**
Closes #146 

### Description

Implement Sign and Verify prompts.

_Note: I'm opening this PR for early feedback. The UI can be improved and the code cleaned up a bit more._

### Steps to Test

**Sign**
1. Open the browser devtools console
2. Enter `webln.enable()` and click the Confirm button on the popup
3. Enter `webln.signMessage('hello').then(res => console.log(res))` 
4. In the Joule popup, verify the message `hello` is displayed
5. Click the `Confirm` button
6. Verify that the generated signature is displayed in the console log

**Verify**
1. Open the browser devtools console
2. Enter `webln.enable()` and click the Confirm button on the popup
3. Enter `webln.verifyMessage('<signature>', 'hello').then(res => console.log(res))` replacing `<signature>` with the output from the `signMessage` call above
4. In the Joule popup, verify the message and signature are displayed correctly
5. Click the `Verify` button
6. Confirm that you receive a success message

### Screenshots

image

image

image

image

### CODE REVIEW

**Comment by jamaljsr** [comment]
@wbobeirne This is great feedback. I really appreciate you taking the time to review. All of your requests are clear to me. I will use the new UI as a base and make the changes you suggested. I don't think it'll take too long, so I should have it done in the next day or two.

On a side note, regarding the signing data structure. I agree that we should only accept JSON as the message argument to `verifyMessage`. Should I include this functionality in this PR or should we keep it separate for now. My concern is releasing this as is then having to worry about backward compatibility when we move to JSON.

File: src/app/lib/lnd-http/index.ts [hunk 1]
```diff
@@ -324,6 +324,26 @@ export class LndHttpClient {
...
+    const msg = new Buffer(message).toString('base64');
+    return this.request<T.SignMessageResponse, T.SignMessageParams>(
+      'POST',
+      '/v1/signmessage',
+      { msg },
+    );
+  };
+
+  verifyMessage = (params: T.VerifyMessageParams) => {
+    return this.request<T.VerifyMessageResponse, T.VerifyMessageParams>(
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/lib/lnd-http/index.ts]: This code is actually just adding the type for the LND REST API response. The WebLN `verifyMessage` function in `provider.ts` still returns void.

File: src/app/prompts/verify.tsx [hunk 2]
```diff
@@ -1,12 +1,174 @@
...
+              <div className="VerifyPrompt-graphic-icon">
+                <img src={this.origin.icon || ''} />
+              </div>
+              <div className="VerifyPrompt-graphic-divider">
+                <Icon type="swap" />
+              </div>
+              <div className="VerifyPrompt-graphic-icon">
+                <img src={Logo} />
+              </div>
+            </div>
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/prompts/verify.tsx]: fixed

File: src/app/prompts/verify.tsx [hunk 3]
```diff
@@ -1,12 +1,174 @@
...
+            <h2 className="VerifyPrompt-title">
+              <strong>{this.origin.name}</strong>
+              requests your verification
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/prompts/verify.tsx]: Updated the message in the UI to just say "Signature Verification"

File: src/app/prompts/verify.tsx [hunk 4]
```diff
@@ -1,12 +1,174 @@
...
+            </Button>
+        }
+        </div>
       </PromptTemplate>
     );
   }
+
+  private handleConfirm = () => {
+    this.setState({ isVerifying: true }, () => {
+      this.props.verifyMessage(this.signature, this.msg);
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/prompts/verify.tsx]: Fixed

File: src/app/prompts/sign.less [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        height: 100%;
+      }
+    }
+
+    &-divider {
+      padding: 0 1rem;
+      font-size: 1.4rem;
+      opacity: 0.4;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/prompts/sign.less]: I refactored this into a separate component. I also reused the node icon/pubkey/alias from the payment prompt, so I refactored that into a shared component as well.

**Comment by jamaljsr** [comment]
I thought about the feedback you gave and took the verification screen a bit further, to make the results stand out a lot more. Check the screenshots below and let me know if I went too far. 😄 

image

image

image

**Comment by wbobeirne** [comment]
Hmm, I actually think in the case of a verified signature, it's important to show the message without any additional clicks needed. If someone were trying to be deceptive, they might change out the message contents, and unobservant users might just see "Verified" and assume it's what they saw on the site. This is definitely a major edge-case, but I think verification from Joule should be reasonably trustable at a glance, and not require a deeper dive.

I think you could still fit in the two tabs and the textareas if you just made verified / invalid message one line with the icon to the left.

**Comment by jamaljsr** [comment]
Ah, you’re right. Good catch.

I’ll make the change later on today.

**Pull request merged**